rust/lance/src/dataset/write/commit.rs

```diff
@@ -542,6 +556,117 @@ mod tests {
         assert_eq!(writes, 2);
     }
 
+    #[tokio::test]
+    async fn test_commit_iops() {
+        // If there's no conflicts, we should be able to commit in 2 io requests:
+        // * write txn file (this could be optional one day)
+        // * write manifest
+        let session = Arc::new(Session::default());
+        let io_tracker = Arc::new(StatsHolder::default());
+        let write_params = WriteParams {
+            store_params: Some(ObjectStoreParams {
+                object_store_wrapper: Some(io_tracker.clone()),
+                ..Default::default()
+            }),
+            session: Some(session.clone()),
+            ..Default::default()
+        };
+        let data = RecordBatch::try_new(
+            Arc::new(ArrowSchema::new(vec![ArrowField::new(
+                "a",
+                DataType::Int32,
+                false,
+            )])),
+            vec![Arc::new(Int32Array::from(vec![0; 5]))],
+        )
+        .unwrap();
+        let dataset = InsertBuilder::new("memory://")
+            .with_params(&write_params)
+            .execute(vec![data])
+            .await
+            .unwrap();
+
+        io_tracker.incremental_stats(); // Reset the stats
+        let read_version = dataset.manifest().version;
+        let _ = CommitBuilder::new(Arc::new(dataset))
+            .execute(sample_transaction(read_version))
+            .await
+            .unwrap();
+
+        // Assert io requetss
+        let io_stats = io_tracker.incremental_stats();
+        assert_eq!(io_stats.read_iops, 0);
```

Comment:
To accomplish this, we'll need to move the call to `checkout_latest()` to the end of the retry loop. This makes the first commit attempt blind, but I think that's desirable as it makes the case of sequential writes faster, without making concurrent writes much slower.

Reply:
I tried that and it worked, but I ended up reverting this change, because by doing that we will let the commit succeed in conflict resolution and eventually write transaction file and manifest file and then realize the version is conflict. That means it makes the retry path slower with 2 more writes. I feel it is not worth the effort, but I might have missed some good way to achieve both.